File: seminar-5/stack/cpp/stack_by_array.cpp
#include <stdexcept>
#include <iostream>

template<typename T>
class Stack {
private:
    T* data;     // динамический массив
    int size_;   // текущее количество элементов
    int capacity;

    /**
     * Увеличивает ёмкость массива.
     */
    void resize(int newCapacity) {
        T* newData = new T[newCapacity];
        for (int i = 0; i < size_; i++) {
            newData[i] = data[i];
        }
        delete[] data;
        data = newData;
        capacity = newCapacity;
    }

public:
    /**
     * Конструктор.
     */
    Stack(int initialCapacity = 4)
        : size_(0), capacity(initialCapacity) {
        data = new T[capacity];
    }

    /**
     * Деструктор.
     */
    ~Stack() {
        delete[] data;
    }

    /**
     * Добавляет элемент на вершину стека.
     */
    void push(const T& value) {
        if (size_ == capacity) {
            resize(capacity * 2);
        }
        data[size_] = value;
        size_++;
    }

    /**
     * Удаляет и возвращает верхний элемент.
     */
    T pop() {
        if (isEmpty()) {
            throw std::runtime_error("pop from empty stack");
        }
        size_--;
        return data[size_];
    }

    /**
     * Возвращает верхний элемент без удаления.
     */
    T peek() const {
        if (isEmpty()) {
            throw std::runtime_error("peek from empty stack");
        }
        return data[size_ - 1];
    }

    /**
     * Проверяет, пуст ли стек.
     */
    bool isEmpty() const {
        return size_ == 0;
    }

    /**
     * Возвращает размер стека.
     */
    int size() const {
        return size_;
    }
};

File: seminar-05/stack/java/StackByArray.java
public class StackByArray<T> {
    private Object[] data; // внутренний массив
    private int size;      // текущее количество элементов

    /**
     * Создаёт пустой стек.
     */
    public StackByArray(int capacity) {
        data = new Object[capacity];
        size = 0;
    }

    public StackByArray() {
        this(4);
    }

    /**
     * Увеличивает ёмкость массива.
     */
    private void resize(int newCapacity) {
        Object[] newData = new Object[newCapacity];
        for (int i = 0; i < size; i++) {
            newData[i] = data[i];
        }
        data = newData;
    }

    /**
     * Добавляет элемент на вершину стека.
     */
    public void push(T value) {
        if (size == data.length) {
            resize(data.length * 2);
        }
        data[size] = value;
        size++;
    }

    /**
     * Удаляет и возвращает верхний элемент.
     */
    @SuppressWarnings("unchecked")
    public T pop() {
        if (isEmpty()) {
            throw new IllegalStateException("pop from empty stack");
        }
        size--;
        T value = (T) data[size];
        data[size] = null; // очищаем ссылку
        return value;
    }

    /**
     * Возвращает верхний элемент без удаления.
     */
    @SuppressWarnings("unchecked")
    public T peek() {
        if (isEmpty()) {
            throw new IllegalStateException("peek from empty stack");
        }
        return (T) data[size - 1];
    }

    /**
     * Проверяет, пуст ли стек.
     */
    public boolean isEmpty() {
        return size == 0;
    }

    /**
     * Возвращает размер стека.
     */
    public int size() {
        return size;
    }

    @Override
    public String toString() {
        Object top = isEmpty() ? null : data[size - 1];
        return "Stack(size=" + size + ", top=" + top + ")";
    }
}
